mathics/builtin/numeric.py

```diff
@@ -62,8 +65,107 @@ def log_n_b(py_n, py_b) -> int:
     return int(mpmath.ceil(mpmath.log(py_n, py_b))) if py_n != 0 and py_n != 1 else 1
 
 
-def apply_N(expression, evaluation, prec=SymbolMachinePrecision):
-    return Expression("N", expression, prec).evaluate(evaluation)
+def apply_N(expr, evaluation, prec=SymbolMachinePrecision):
+    """
+    This function applies recursively `N-value` rules for convert symbols, expressions and exact numbers
+    into Real numbers with certain precision.
+    Notice that `apply_N(expr, evaluation, prec)` is not equivalent in general to call
+    `Expression(SymbolN, expr, prec).evaluate(evaluation)`. To get the same behaviour, in general
+    is necesary to call
+    `apply_N(expr.evaluate(evaluation), evaluation, prec.evaluate(evaluation)).evaluate(evaluation)`
+    However, in many cases where this function is called from Python code, both `prec` and `expr`
+    are given in an already evaluated form, while the last evaluation is only required when
+    `apply_N` failed to return a real number.
+    """
+    try:
+        d = get_precision(prec, evaluation)
+    except PrecisionValueError:
+        return
+
+    if isinstance(expr, Number):
+        return expr.round(d)
+    if isinstance(expr, String):
+        return expr
```

Comment:
Why accept strings here?

Reply:
Imagine you have the expression

`N[MyFunc[Pi^2, Method->"mymethod"]]`

Then, at some point, you have to apply `apply_N("mymethod")`. But then, you do not want to look for a pattern that converts "mymethod" to a number. Then, the best is to return the expression immediately.

Reply:
Oh, I understand, `apply_N` is recursive. Thanks!

Reply:
Yep. This is why I can not put the conditional evaluation *inside* `apply_N`.